common/scala/src/main/scala/whisk/core/entity/FullyQualifiedEntityName.scala

```diff
@@ -31,11 +31,24 @@ import whisk.core.entity.size.SizeString
  */
 protected[core] case class FullyQualifiedEntityName(path: EntityPath, name: EntityName, version: Option[SemVer] = None) extends ByteSizeable {
     private val qualifiedName: String = path + EntityPath.PATHSEP + name
+    /** Resolves default namespace in path to given name if the root path is the default namespace. */
     def resolve(namespace: EntityName) = FullyQualifiedEntityName(path.resolveNamespace(namespace), name, version)
+
+    /** @return full path including name, i.e., "path/name" */
+    def fullPath: EntityPath = path.addPath(name)
+
+    /**
+     * Creates new fully qualified entity name that shifts the name into the path and adds a new name:
+     * (p, n).add(x) -> (p/n, x).
+     *
+     * @return new fully qualified name
+     */
+    def add(n: EntityName) = FullyQualifiedEntityName(path.addPath(name), n)
+
     def toDocId = DocId(qualifiedName)
-    def pathToDocId = DocId(path())
     def qualifiedNameWithLeadingSlash: String = EntityPath.PATHSEP + qualifiedName
-    def apply() = path.addpath(name) + version.map("@" + _.toString).getOrElse("")
+    def apply() = path.addPath(name) + version.map("@" + _.toString).getOrElse("")
```

Comment:
Not a fan of `apply()` methods in general, they're not particularly informative about what they do (sorry that I missed that before). I'd much, much, rather see: `fqn.expandedName` that `fqn()`.